cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/CustomActionExecutor.java

```diff
@@ -0,0 +1,183 @@
+   * @param classLoader used to load the custom action class
+   * @throws Exception when failed to instantiate the custom action
+   */
+  public CustomActionExecutor(ProgramRunId workflowRunId, BasicWorkflowContext workflowContext,
+                              InstantiatorFactory instantiator, ClassLoader classLoader) throws Exception {
+    this.workflowRunId = workflowRunId;
+    this.workflowContext = workflowContext;
+    this.action = createAction(workflowContext, instantiator, classLoader);
+  }
+
+  @SuppressWarnings("unchecked")
+  private WorkflowAction createAction(BasicWorkflowContext context, InstantiatorFactory instantiator,
+                                      ClassLoader classLoader) throws Exception {
+    Class<?> clz = Class.forName(context.getSpecification().getClassName(), true, classLoader);
+    Preconditions.checkArgument(WorkflowAction.class.isAssignableFrom(clz), "%s is not a WorkflowAction.", clz);
+    WorkflowAction action = instantiator.get(TypeToken.of((Class<? extends WorkflowAction>) clz)).create();
+    Reflections.visit(action, action.getClass(),
+                      new PropertyFieldSetter(context.getSpecification().getProperties()),
+                      new DataSetFieldSetter(context),
+                      new MetricsFieldSetter(context.getMetrics()));
+    return action;
+  }
+
+  void execute() throws Exception {
+    CountDownLatch executorTerminateLatch = new CountDownLatch(1);
+    ExecutorService executorService = createExecutor(1, executorTerminateLatch, "action-"
+      + workflowContext.getSpecification().getName() + "-%d");
+
+    try {
+      // Run the action in new thread
+      Future<?> future = executorService.submit(new Callable<Void>() {
+        @Override
+        public Void call() throws Exception {
+          setContextCombinedClassLoader(action);
+          try {
+            initializeInTransaction(action, workflowContext);
+            runInTransaction(action, workflowContext);
+            return null;
+          } finally {
+            destroyInTransaction(action, workflowContext);
+          }
+        }
+      });
+      future.get();
+    } catch (Throwable t) {
+      LOG.error("Error executing the action {} in the Workflow {}.", workflowContext.getSpecification().getName(),
+                workflowRunId, t);
+      Throwables.propagateIfPossible(t, Exception.class);
+      throw Throwables.propagate(t);
+    } finally {
+      executorService.shutdownNow();
+      executorTerminateLatch.await();
+    }
+  }
+
+  private void initializeInTransaction(WorkflowAction action, BasicWorkflowContext context) throws Exception {
+    TransactionContext transactionContext = context.getDatasetCache().newTransactionContext();
+    transactionContext.start();
+    try {
+      action.initialize(context);
+    } catch (Throwable t) {
+      LOG.error("Failed to initialize the Workflow action {} for Workflow run {}.",
```

Comment:
Will it result in logging too much? The `transactionContext.abort` will throw exception and the caller (WorkflowDriver) will catch and log again?